agent/php_call.c

```diff
@@ -48,53 +49,91 @@ zval* nr_php_call_user_func(zval* object_ptr,
     }
   }
 
-  retval = nr_php_zval_alloc();
-
   fname = nr_php_zval_alloc();
   nr_php_zval_str(fname, function_name);
-#if ZEND_MODULE_API_NO >= ZEND_8_2_X_API_NO /* PHP 8.2+ */
+
+  /*
+   * For PHP 8+, in the case of exceptions according to:
+   * https://www.php.net/manual/en/function.call-user-func.php
+   * Callbacks registered with functions such as call_user_func() and
+   * call_user_func_array() will not be called if there is an uncaught exception
+   * thrown in a previous callback. So if we call something that causes an
+   * exception, it will block us from future calls that use call_user_func or
+   * call_user_func_array.
+   */
+
+#if ZEND_MODULE_API_NO >= ZEND_8_2_X_API_NO /* PHP 8.2+ : if clause 2*/
   /*
    * With PHP 8.2, functions that do not exist will cause a fatal error to
    * be thrown. `zend_call_method_if_exists` will attempt to call a function and
-   * silently fail if it does not exist
+   * silently fail if it does not exist.  This calls call_user_function under
+   * the hood.
    */
-  if (NULL != object_ptr) {
-    object = Z_OBJ_P(object_ptr);
-  } else {
-    object = NULL;
-  }
 
-  if (NULL != fname) {
-    method_name = Z_STR_P(fname);
-  } else {
-    return NULL;
-  }
   zend_try {
-  zend_result = zend_call_method_if_exists(object, method_name, retval,
-                                           param_count, param_values);
+    if (NULL != object_ptr) {
+      object = Z_OBJ_P(object_ptr);
+    } else {
+      object = NULL;
+    }
+
+    if (NULL != fname) {
+      method_name = Z_STR_P(fname);
+    } else {
+      return NULL;
+    }
+
+    retval = nr_php_zval_alloc();
+    zend_result = zend_call_method_if_exists(object, method_name, retval,
+                                             param_count, param_values);
+  }
+  zend_catch {
+    zend_result = FAILURE;
+  }
+  nr_php_zval_free(&fname);
+
+  nr_free(param_values);
+  if (SUCCESS == zend_result) {
+    return retval;
   }
-  zend_catch { zend_result = FAILURE; }
+  nr_php_zval_free(&retval);
+  return NULL;
   zend_end_try();
```

Comment:
I don't know a whole lot about zend_try/zend_catch/zend_end_try but intuitively the order of these lines does not seem right. I.e. returning from function before `zend_end_try`.

Reply:
good point.  please see: [d818132](https://github.com/newrelic/newrelic-php-agent/pull/638/commits/d8181329c518cfe29489eb30e4141afe978bd745)

Reply:
Resolved with #640